### PR TITLE
euler discrete set_timesteps() bug fix

### PR DESCRIPTION
**Pull request by edgarmkhitaryan**
In EulerDiscreteScheduler when setting the timesteps we need to exclude the 0-th timestep. So, due to this bug during the inference the scheduler expects to receive x0 and its corresponding noise in order to denoise it which is a redundant operation because we assume that  x0 is our final result and there is no need to remove noise from it. 

After the bug fix, improvements are observed in the quality of the generated images.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @edgarmkhitaryan,

Could you add an example that clearly shows how the current approach is incorrect and an example of how your fix improves it? Thanks!

**Comment by patrickvonplaten** [comment]
Thank you! 

Can you send a link to a paper here that shows how the current implementation is incorrect? I've ran both implementation on a bunch of examples and I honestly don't think one or the other is better.

The problem here is that this change would break backwards compatibility for all schedulers, so I'm not too keen on adding it just like this. 

Generally, we don't really work in timestep space, but rather in "sigma" space: https://github.com/huggingface/diffusers/blob/e5215dee9acbe5d6105f0c40744c4fd676f770bf/src/diffusers/schedulers/scheduling_euler_discrete.py#L175
so every timestep is always converted to sigmas: https://github.com/huggingface/diffusers/blob/e5215dee9acbe5d6105f0c40744c4fd676f770bf/src/diffusers/schedulers/scheduling_euler_discrete.py#L315

**However** what we could do here to allow for more customization, is to just allow the user to pass `timesteps` to https://github.com/huggingface/diffusers/blob/e5215dee9acbe5d6105f0c40744c4fd676f770bf/src/diffusers/schedulers/scheduling_euler_discrete.py#L182 and we then change:

```py
timesteps = np.linspace(0, self.config.num_train_timesteps - 1, num_inference_steps, dtype=float)[::-1].copy()
```

to

```py
if timesteps == None:
    timesteps = np.linspace(0, self.config.num_train_timesteps - 1, num_inference_steps, dtype=float)[::-1].copy()
```

E.g. you can see how it is implemented for DDPM and maybe copy this logic to euler discrete?
https://github.com/huggingface/diffusers/blob/e5215dee9acbe5d6105f0c40744c4fd676f770bf/src/diffusers/schedulers/scheduling_ddpm.py#L185

**Comment by edgarmkhitaryan** [comment]
In my previous comment I forgot to mention that the difference can be visible when using smaller number of timesteps (e.g. in the pervious comment's example the number of timesteps was 10).


In any paper in the topic of diffusion models you can clearly see that x[0] is the final denoised image, so there is no need to one more time remove noise from it.

Let me explain in more intuitive way why this current implementation is incorrect.

imagine we want to do the inference with EulerDiscreteScheduler with 5 timesteps.


```
timesteps = [999.  , 749.25, 499.5 , 249.75,   0. ]

x[t] ~ Gaussian
for t in timesteps:
    
    noise_t = unet(x[t], t)
    
    x[t-1] = scheduler.step(x[t], noise_t, t)
```


The problem appears when the algorithms reaches to t=0

```
noise_t = unet(x[0], 0)

x[t-1] = scheduler.step(x[0], noise_t, 0)
```

So, if our model is trained in such a way that x[0] is the final image without any noise, why do we try one more time to detect the noise in the x[0] and remove the noise from it? 

The last step is redundant for sure!

The solution that I offer is just to exclude the 0-th timestep. In that case we would get these timesteps for the inference.


```
timesteps = [999. , 799.2, 599.4, 399.6, 199.8]
```

In this case out 5-th step won't be redundant. In addition, in this way the difference between the timesteps get smaller which also will boost the performance as well (e.g. from 999-th step it is much easier for the model to get 799.2-th result compared to 749.25-th).


When using bigger number of timesteps the generated images are nearly identical, the bug is visible only when using small number of timesteps. 

This is clearly a bug and needs to be fixed.
